File: front-end/src/services/BbService.js
// BbService.js
// 목적 : b 정보를 axios 로 CRUD 를 하는 공통 함수들 정의
import http from "../http-common";

class BbService {
  // 전체 조회 (페이징)
  getAll(bb_ctitle, bb_title, page, size) {
    return http.get(
      `/bb?bb_ctitle=${bb_ctitle}&bb_title=${bb_title}&page=${page}&size=${size}`,
      {}
    );
  }
  
  // 조회수 best 5
  getBbBoardBest() {
    return http.get(`/bb/best`);
  }

  // 댓글수 best 5 
  getBbBoardRcntBest() {
    return http.get(`/bb/rcnt`);
  }

   // 최신순 5
   getBbDesc() {
    return http.get(`/bb/desc`);
  }

  // 상세페이지 조회
  getDetailAll() {
    return http.get(`/bb/all`);
  }

  // 게시번호로 조회
  get(bno) {
    return http.get(`/bb/${bno}`);
  }

  // 게시글 생성
  create(nick, bbCategory, bbTitle, bbContent, bbTag, blobFile, onUploadProgress) {
    let formData = new FormData();

    console.log(nick);
    console.log(bbCategory);
    console.log(bbTitle);
    console.log(bbContent);
    console.log(bbTag);

    console.log(blobFile);
    
    formData.append("nick", nick);
    formData.append("bbCategory", bbCategory);
    formData.append("bbTitle", bbTitle);
    formData.append("bbContent", bbContent);
    formData.append("bbTag", bbTag);
    formData.append("blobFile", blobFile);

    return http.post("/bb", formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
      onUploadProgress
    });
  }
  
  // 수정
  update(
    bno,
    nick,  
    bbCategory,
    bbTitle,
    bbContent,
    bbTag,
    bbCnt,
    blobFile,
    onUploadProgress
  ) {
    let formData = new FormData();

    console.log(bno);
    console.log(nick);
    console.log(bbCategory);
    console.log(bbTitle);
    console.log(bbContent);
    console.log(bbTag);
    console.log(bbCnt);

    console.log(blobFile);

    formData.append("bno", bno);
    formData.append("nick", nick);
    formData.append("bbCategory", bbCategory);
    formData.append("bbTitle", bbTitle);
    formData.append("bbContent", bbContent);
    formData.append("bbTag", bbTag);
    formData.append("bbCnt", bbCnt);
    formData.append("blobFile", blobFile);

    return http.put(`/bb/${bno}`, formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
      onUploadProgress,
    });
  }

  // 삭제
  delete(bno) {
    return http.delete(`/bb/deletion/${bno}`);
  }
}

export default new BbService();

File: DPD_backend/src/main/java/com/example/simpledms/service/community/BbService.java
package com.example.simpledms.service.community;

import com.example.simpledms.dto.community.BbCategoryDto;
import com.example.simpledms.dto.community.BbCommentDto;
import com.example.simpledms.model.community.Bb;
import com.example.simpledms.repository.community.BbRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;
import java.util.Optional;

/**
 * packageName : com.example.simpledms.service.community
 * fileName : BbService
 * author : ds
 * date : 2022-12-19
 * description : 추천게시판 서비스 클래스
 * ===========================================================
 * DATE            AUTHOR             NOTE
 * —————————————————————————————
 * 2022-12-19         ds          최초 생성
 */
@Service
public class BbService {
    // CRUD 함수 사용을 위한 Repository 불러오기
    @Autowired
    BbRepository bbRepository;

    // 이미지 포함한 생성 메서드
    public Bb createUploadFile(String nick,
                               String bbCategory,
                               String bbTitle,
                               String bbContent,
                               String bbTag,
                               MultipartFile blobFile)
            throws IOException {

        if(blobFile != null) {
            // 업로드 파일에서 파일명 얻기
            String fileName = StringUtils.cleanPath(blobFile.getOriginalFilename());

            Bb bb = Bb.builder()
                    .nick(nick)
                    .bbCategory(bbCategory)
                    .bbTitle(bbTitle)
                    .bbContent(bbContent)
                    .bbTag(bbTag)
                    .fileName(fileName)
                    .blobFile(blobFile.getBytes())
                    .build();

            Bb createBb = bbRepository.save(bb);
            return createBb;

        } else {
            Bb bb = Bb.builder()
                    .nick(nick)
                    .bbCategory(bbCategory)
                    .bbTitle(bbTitle)
                    .bbContent(bbContent)
                    .bbTag(bbTag)
                    .fileName("")
                    .blobFile(null)
                    .build();

            Bb createBb = bbRepository.save(bb);
            return createBb;
        }
    }

    // 이미지 포함한 수정 메서드
    public Bb updateUploadFile(int bno, String nick, String bbCategory, String bbTitle, String bbContent, String bbTag, int bbCnt,
                               MultipartFile blobFile) throws IOException {

        if (blobFile != null) {
            // 업로드 파일에서 파일명 얻기
            String fileName = StringUtils.cleanPath(blobFile.getOriginalFilename());

            Bb bb = Bb.builder()
                    .bno(bno)
                    .nick(nick)
                    .bbCategory(bbCategory)
                    .bbTitle(bbTitle)
                    .bbContent(bbContent)
                    .bbTag(bbTag)
                    .bbCnt(bbCnt)
                    .blobFile(blobFile.getBytes())
                    .build();

            Bb createBb = bbRepository.save(bb);
            return createBb;
        } else {
            Bb bb = Bb.builder()
                    .bno(bno)
                    .nick(nick)
                    .bbCategory(bbCategory)
                    .bbTitle(bbTitle)
                    .bbContent(bbContent)
                    .bbTag(bbTag)
                    .bbCnt(bbCnt)
                    .blobFile(null)
                    .build();

            Bb createBb = bbRepository.save(bb);
            return createBb;
        }
    }

    // 카테고리 선택 후 검색 조회 메서드
    public Page<BbCategoryDto> selectSearchAndKeywordSearch(String bb_ctitle,
                                                            String bb_title,
                                                            Pageable pageable) {
        Page<BbCategoryDto> bbPage
                = bbRepository.selectSearchAndKeywordSearch(bb_ctitle, bb_title, pageable);

        return bbPage;
    }

    // 자유게시판 상세 게시판 메소드
    public List<Bb> findAll() {

        List<Bb> bbList = bbRepository.findAll();
        return bbList;
    }

    // 게시판 번호(bno) 로 상세조회 메서드
    public Optional<Bb> findId(int bno) {

        // findById : parameter 값 - ID, return  값 - Optional
        Optional<Bb> bbOptional = bbRepository.findById(bno);

        return bbOptional;
    }

    // 게시판 번호(bno)로 삭제하는 메서드
    public boolean removeById(int bno) {

        if (bbRepository.existsById(bno)) {
            bbRepository.deleteById(bno);
            return true;
        }
        return false;
    }

    // 상세페이지 클릭시 조회수 1상승 메서드
    // @Transactional : 해당 범위 내 메서드가 트랜잭션이 되도록 보장해준다.
    @Transactional
    public int updateCnt(Integer bno) {
        return bbRepository.updateCnt(bno);
    }

    // 조회수 Best 5 리스트 메서드
    public List<Bb> boardBestList() {
        List<Bb> list = bbRepository.boardBestList();

        return list;
    }

    // 최신순 6 리스트 메서드
    public List<Bb> boardDescList() {
        List<Bb> list = bbRepository.boardDescList();

        return list;
    }

    // 댓글수 Best 5 리스트 메서드
    public List<BbCommentDto> findBestRcnt() {
        List<BbCommentDto> list = bbRepository.findBestRcnt();

        return list;
    }
}
